### ISSUE TITLE
Opening the flutter/engine repo results in errors

### ISSUE THREAD

**Issue by dnfield** [Dart-Code/Dart-Code#692]
# Problem:
Opening https://github.com/flutter/engine repo in VSCode results in many errors

## Steps to reproduce
Open the engine repo
- By default, uses the first Dart SDK on Path instead of Flutter's Dart
- Adding a Flutter folder to the workspace still results in errors (using Flutter's version of dart)
- Running `flutter packages get` in folders that have a `pubspec.yaml` resolves some errors but still leaves many others

Screenshots:
With just engine (note it's using Dart 1.24.3)
![Without another workspace](https://user-images.githubusercontent.com/8620741/37422904-cdd19810-2792-11e8-93a6-2aec447d2064.png)

Adding another flutter project to workspace:
![With a flutter project folder at root of workspace](https://user-images.githubusercontent.com/8620741/37423010-20fda484-2793-11e8-9638-44733b512fcf.png)

**Comment by DanTup** [comment]
I cloned the repo using `gclient`, though my VS install is bust so I couldn't "build" anything. I think that might be the cause of some of these issues, so I'll have to try again when I have my MacBook up and running.

Seems like there are a few issues though:

## 1. We don't detect that this repo is a "flutter project"

To fix this, we need a way of detecting this tree. Normally to detect Flutter we look in `pubspec.yaml` and see if it contains `sdk: flutter` though we also have a special case for if the folder contains a `bin/flutter` file and `bin/cache-sdk` folder for the flutter/flutter repo. I can't see anything obvious we could use here (that is specific enough to not cause false-positives).

## 2. If we do detect it as a Flutter project, we don't know where to find the SDK

You can fix this by setting `FLUTTER_ROOT` or when it pops up saying it couldn't find a Flutter SDK, browsing to one with the `Locate SDK` button (which will save to your global VS Code settings). You could also then more it to your workspace settings.

If running the build properly creates its own SDK, then we should add a path for it to the search paths, so that we use its own built SDKs (I don't know if this is the case - it seems weird for flutter/flutter to depend on flutter/engine and flutter/engine to depend on flutter/flutter!?).

## 3. Missing packages

Many of these are solved by running `flutter packages get` in each folder, however some of them (like the examples folder) have empty pubspecs. I don't know how these packages are supposed to be fetched. Possibly this would be fixed by building properly (see note above).

## Missing libraries

There is code in here that references libraries I don't know of (eg. `dart:mozart.internal` in `./lib/snapshot/snapshot_fuchsia.dart`). Again, I don't know where these should be coming from. Possibly this would be fixed by building properly (see note above).

**Comment by DanTup** [comment]
So, for now:

1. You can workaround this by adding another flutter project as a workspace root (eg. flutter/flutter's examples/hello_world)
2. You can workaround this by seting `dart.flutterSdkPath` or the env var `FLUTTER_ROOT` to point at appropriate folders
3. This one needs some digging - how is this supposed to work?
4. This one needs some digging - where are these supposed to come from?

If I get my MacBook sorted next week, I'll have a quick play to see if I can build and figure out the others - if you figure out answers to any of them, let me know.

**Comment by DanTup** [comment]
Any idea how this of this works in other IDEs?

**Comment by dnfield** [comment]
Unfortunately no.

**Comment by DanTup** [comment]
I think the missing thing here might be pulling dependencies using gclient:

https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#getting-the-code-and-configuring-your-environment

@dnfield Did you already do this and still have issues?

**Comment by dnfield** [comment]
Yes - I had all the dependencies and was able to run local engine builds

**Comment by dnfield** [comment]
Also - I don't know that I'd prioritize this very highly.  I imagine there are a lot of weird edge cases that might not be easy to solve for in that repo.  Some of the errors get resolved by pulling in necessary packages, and at this point I've learned to tune out the others if/when I'm using VSCode to work on that repo.

**Comment by DanTup** [comment]
Yeah, I wasn't looking at this specifically, I was just cleaning up the issue tracker and came across it so was re-familiarising myself with what was wrong. To progress, I probably need to review my list of issues above and figure out which still apply if I'd actually pulled the dependencies properly.

**Comment by DanTup** [comment]
@dnfield Is this still the still in the same state now as it was then? Can you run the analyzer from the CLI and have it pass?

**Comment by DanTup** [comment]
@dnfield ping!

**Comment by dnfield** [comment]
Haha I have a better setup now but I've also learned to just ignore some of the errors.

**Comment by DanTup** [comment]
Cool, thanks! I don't know what cquery is, but I suspect it may be out of scope for the Dart plugin :-D
